src/clients/QBittorrent.ts

```diff
@@ -215,76 +221,140 @@ export default class QBittorrent implements TorrentClient {
+		}
+		await this.request(
+			"/torrents/recheck",
+			`hashes=${torrent[0].hash}`,
 			X_WWW_FORM_URLENCODED,
 		);
-		try {
-			const properties = JSON.parse(responseText);
-			return properties && typeof properties === "object";
-		} catch (e) {
-			return false;
-		}
 	}
+
+	/*
+	@param searchee the Searchee we are generating off (in client)
+	@return either a string containing the path or a error mesage
+	*/
 	async getDownloadDir(
-		searchee: Searchee,
+		searchee: SearcheeWithInfoHash,
 	): Promise<
 		Result<string, "NOT_FOUND" | "TORRENT_NOT_COMPLETE" | "UNKNOWN_ERROR">
 	> {
-		let torrentInfo: TorrentConfiguration;
 		try {
-			if (await this.isInfoHashInClient(searchee.infoHash!)) {
-				torrentInfo = await this.getTorrentConfiguration(searchee);
-				if (torrentInfo.save_path === undefined) {
-					return resultOfErr("NOT_FOUND");
-				}
+			const result = await this.getTorrentInfo(searchee.infoHash);
+			if (result.length === 0) {
+				return resultOfErr("NOT_FOUND");
 			}
+			const torrentInfo = result[0];
+			if (!this.isTorrentComplete(torrentInfo.state)) {
+				return resultOfErr("TORRENT_NOT_COMPLETE");
+			}
+			const savePath = await this.generateCorrectSavePath(
+				searchee,
+				torrentInfo,
+			);
+			return resultOf(savePath);
 		} catch (e) {
 			if (e.message.includes("retrieve")) {
 				return resultOfErr("NOT_FOUND");
 			}
 			return resultOfErr("UNKNOWN_ERROR");
 		}
-		return resultOf(torrentInfo!.save_path);
 	}
 
-	async getTorrentConfiguration(
+	/*
+	@param searchee the Searchee we are generating off (in client)
+	@param subfolderLayout whether it is subfolder layout or not
+	@return string absolute location from client with content layout considered
+	 */
+	async generateCorrectSavePath(
 		searchee: Searchee,
-	): Promise<TorrentConfiguration> {
-		const responseText = await this.request(
-			"/torrents/info",
-			`hashes=${searchee.infoHash}`,
-			X_WWW_FORM_URLENCODED,
+		torrentInfo: TorrentInfo,
+	): Promise<string> {
+		const subfolderContentLayout = await this.isSubfolderContentLayout(
+			searchee,
+			torrentInfo,
+		);
+		if (subfolderContentLayout) {
+			return dirname(torrentInfo.content_path);
+		}
+		return torrentInfo.save_path;
+	}
+
+	/*
+	@return array of query results
+	 */
+	async getAllTorrentInfo(): Promise<TorrentInfo[]> {
+		const responseText = await this.request("/torrents/info", "");
+		return JSON.parse(responseText);
+	}
+
+	/*
+	@param hash the hash of the torrent or undefined for all torrents
+	@return array of query results
+	 */
+	async getTorrentInfo(hash: string): Promise<TorrentInfo[]> {
+		const torrents = await this.getAllTorrentInfo();
+		return torrents.filter(
+			(torrent) =>
+				hash === torrent.hash ||
+				hash === torrent.infohash_v1 ||
+				hash === torrent.infohash_v2,
 		);
-		const searchResult = JSON.parse(responseText).find(
-			(e) => e.hash === searchee.infoHash,
-		) as TorrentInfo;
-		if (searchResult === undefined) {
+	}
```

Comment:
is this really the only way to handle info hash v2? I see it gets called in several places now, and fetching the entire torrent list every time feels like it'll be slow, especially on large qbit instances (maybe also slower on instances using fastresume files). would it be worth adding an infoHashV2 field into Metafile?

Also, is there ever a situation we expect multiple values in the returned array? If not, we could use `.find` instead of `.filter` and then return `TorrentInfo | undefined`

Reply:
> is this really the only way to handle info hash v2?

Yes, if we use the v1 infohash, it won't return anything. The only other solution is like you said to calculate v2. But first I'll try to see if I can pass around torrentInfo so we can only call it once.

> Also, is there ever a situation we expect multiple values in the returned array?

I don't think there would be multiple values. qbit uses `torrent.hash` for everything and that can't be duplicated. `torrent.infohash_v2` is more characters than v1 (but gets tructated to 40 for `torrent.hash`) so those would never be duplicated.

Reply:
See: https://github.com/cross-seed/cross-seed/pull/652/files#r1586504034

Reply:
It shouldn't be hard to also keep track of v2 info hashes. But this PR is already big enough, I think this solution is fine for now.